### DISCUSSION TITLE
Speed up coroutine creation ?

### DISCUSSION THREAD

**Discussion by kumaraditya303** [faster-cpython/ideas#191] [Ideas]
Python version -  Python 3.11.0a3+

Script:
```py
import timeit
async def coro():pass
def gen():
    yield

print("Coroutine:",timeit.timeit(lambda:coro(), number=1000000))
print("Generator:",timeit.timeit(lambda:gen(), number=1000000))
```

Results:

```console
Running Debug|x64 interpreter...
<timeit-src>:6: RuntimeWarning: coroutine 'coro' was never awaited
RuntimeWarning: Enable tracemalloc to get the object allocation traceback
Coroutine: 10.316932699999597
Generator: 1.845993200004159
```

Coroutine creation is atleast **5.7x** slower than a generator which is very slow considering that in asyncio coroutine are created at very high rate.

Further investigation revealed that most of the slowness is due to tracking of coroutine if it was awaited or not:

Script:

```py
import timeit
async def coro():pass
def gen():
    yield

print("Coroutine:",timeit.timeit(lambda:coro().close(), number=1000000))
print("Generator:",timeit.timeit(lambda:gen(), number=1000000))
```

Results:

```console 
Coroutine: 1.8880774999997811
Generator: 1.7808137999963947
```

If the created coroutine is closed so that its tracking feature does not track it then it is margin of error slower than generator, revealing that tracking coroutine is slowing it down.

**Comment by gvanrossum** [comment]
Can you test again, adding await or next()?

**Comment by asvetlov** [comment]
Please clarify, did you measure *coroutine creation time* or a *time of raising RuntimeWarning*?

Looing on the script, I suspect the second.
A well-written program never emits such warnings.

**Reply by kumaraditya303** [reply]
Yes it seems like it is more about coroutine tracking than coroutine creation, I'll try to come up with another script which is solely about coroutine creation.
